File: dsa_in_java/linkedlist/PalindromeLinkedList.java
/*
 * Topic: Palindrome Linked List
 * Language: Java
 * 
 * Description:
 * Given the head of a singly linked list, determine if it is a palindrome.
 * 
 * Example:
 * Input: 1 -> 2 -> 2 -> 1
 * Output: true
 * 
 * Approach:
 * 1. Find the middle of the linked list using the slow and fast pointer technique.
 * 2. Reverse the second half of the linked list.
 * 3. Compare the first half and the reversed second half node by node.
 * 4. If all nodes match, the linked list is a palindrome.
 * 5. Restore the original list (optional).
 * 6. Return the result.
 * 
 * Leetcode link:   
 * https://leetcode.com/problems/palindrome-linked-list/
 */

class ListNode {
    int val;
    ListNode next;
    ListNode(int val) { this.val = val; }
}

public class PalindromeLinkedList {

    public static boolean isPalindrome(ListNode head) {
        if (head == null || head.next == null) return true;

        // Find middle
        ListNode slow = head, fast = head;
        while (fast.next != null && fast.next.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        // Reverse second half
        ListNode secondHalf = reverse(slow.next);
        ListNode firstHalf = head;

        // Compare both halves
        while (secondHalf != null) {
            if (firstHalf.val != secondHalf.val) return false;
            firstHalf = firstHalf.next;
            secondHalf = secondHalf.next;
        }

        return true;
    }

    private static ListNode reverse(ListNode head) {
        ListNode prev = null, curr = head;
        while (curr != null) {
            ListNode next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }

    public static void main(String[] args) {
        ListNode head = new ListNode(1);
        head.next = new ListNode(2);
        head.next.next = new ListNode(2);
        head.next.next.next = new ListNode(1);

        System.out.println("Is Palindrome: " + isPalindrome(head));
    }
}

/*
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

File: dsa_in_js/linkedlist/palindrome_linked_list.js
/**
 * Definition for singly-linked list node.
 */
export class ListNode {
	/**
	 * Creates a new ListNode.
	 * @param {number} value The value of the node
	 * @param {ListNode} next The reference to the next node (default is null)
	 */
	constructor(value = 0, next = null) {
		this.value = value;
		this.next = next;
	}
}

/**
 * Checks if a singly linked list is a palindrome using the Two Pointer technique.
 * @param {ListNode} head The head of the linked list
 * @returns {boolean} True if the list is a palindrome, otherwise false
 */
export function isPalindrome(head) {
	if (!head || !head.next) return true;

	// Step 1: Find middle using slow and fast pointers
	let slow = head;
	let fast = head;
	while (fast && fast.next) {
		slow = slow.next;
		fast = fast.next.next;
	}

	// Step 2: Reverse the second half
	let prev = null;
	let curr = slow;
	while (curr) {
		const nextNode = curr.next;
		curr.next = prev;
		prev = curr;
		curr = nextNode;
	}

	// Step 3: Compare first and second halves
	let first = head;
	let second = prev;
	while (second) {
		if (first.value !== second.value) return false;
		first = first.next;
		second = second.next;
	}

	return true;
}

/**
 * Helper function to create a linked list from an array.
 * @param {number[]} arr The array of values
 * @returns {ListNode} The head of the linked list
 */
export function createLinkedList(arr) {
	if (arr.length === 0) return null;
	const head = new ListNode(arr[0]);
	let current = head;
	for (let i = 1; i < arr.length; i++) {
		current.next = new ListNode(arr[i]);
		current = current.next;
	}
	return head;
}

/**
 * Helper function to print the linked list (for testing purposes).
 * @param {ListNode} head The head of the linked list
 */
export function printList(head) {
	const values = [];
	let current = head;
	while (current) {
		values.push(current.value);
		current = current.next;
	}
	console.log(values.join(' -> '));
}

/**
 * Tiny self-check in browser console.
 */
if (typeof window !== "undefined")
(function selfTest() {
	const head1 = createLinkedList([1, 2, 3, 2, 1]);
	console.log("List 1:");
	printList(head1);
	console.log("Is palindrome? ->", isPalindrome(head1));

	console.log("---");

	const head2 = createLinkedList([1, 2, 3, 4]);
	console.log("List 2:");
	printList(head2);
	console.log("Is palindrome? ->", isPalindrome(head2));
})();
